mlflow/resource_experiment.go

```diff
@@ -0,0 +1,111 @@
+	Description      string `json:"description,omitempty"`
+	ExperimentId     string `json:"experiment_id,omitempty" tf:"computed"`
+	ArtifactLocation string `json:"artifact_location,omitempty" tf:"force_new"`
+	LifecycleStage   string `json:"lifecycle_stage,omitempty" tf:"computed"`
+	LastUpdateTime   int64  `json:"last_update_time,omitempty" tf:"computed"`
+	CreationTime     int64  `json:"creation_time,omitempty" tf:"computed"`
+}
+
+type experimentUpdate struct {
+	ExperimentId string `json:"experiment_id"`
+	NewName      string `json:"new_name"`
+}
+
+type experimentWrapper struct {
+	Experiment Experiment `json:"experiment"`
+}
+
+// ExperimentsAPI ...
+type ExperimentsAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+// NewExperimentsAPI ...
+func NewExperimentsAPI(ctx context.Context, m interface{}) ExperimentsAPI {
+	return ExperimentsAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// Create ...
+func (a ExperimentsAPI) Create(d *Experiment) error {
+	return a.client.Post(a.context, "/mlflow/experiments/create", d, &d)
+}
+
+// Read ...
+func (a ExperimentsAPI) Read(experimentId string) (*Experiment, error) {
+	var d experimentWrapper
+	err := a.client.Get(a.context, "/mlflow/experiments/get", map[string]string{
+		"experiment_id": experimentId,
+	}, &d)
+	if err != nil {
+		return nil, err
+	}
+	return &d.Experiment, nil
+}
+
+// Update ...
+func (a ExperimentsAPI) Update(e *experimentUpdate) error {
+	return a.client.Post(a.context, "/mlflow/experiments/update", e, &e)
+}
+
+// Delete ...
+func (a ExperimentsAPI) Delete(e *Experiment) error {
+	return a.client.Post(a.context, "/mlflow/experiments/delete", e, &e)
+}
+
+func ResourceMLFlowExperiment() *schema.Resource {
+	s := common.StructToSchema(
+		Experiment{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+			return m
+		})
+
+	return common.Resource{
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var e Experiment
+			if err := common.DataToStructPointer(d, s, &e); err != nil {
+				return err
+			}
+			if err := NewExperimentsAPI(ctx, c).Create(&e); err != nil {
+				return err
+			}
+			d.SetId(e.ExperimentId)
+			return nil
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			e, err := NewExperimentsAPI(ctx, c).Read(d.Id())
+			if err != nil {
+				return err
+			}
+			return common.StructToData(*e, s, d)
+		},
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var e Experiment
+			if err := common.DataToStructPointer(d, s, &e); err != nil {
+				return err
+			}
+			updateDoc := experimentUpdate{ExperimentId: d.Id(), NewName: e.Name}
+			return NewExperimentsAPI(ctx, c).Update(&updateDoc)
+		},
+		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			e := Experiment{ExperimentId: d.Id()}
+			return NewExperimentsAPI(ctx, c).Delete(&e)
```

Comment:
Can you refactor Delete method to accept just a string id? E.g. we don't need any other fields than ID to delete an entity. And within that method you'll probably use that struct, but outer interface would be simple and consistent with the rest of resources.

Consistency is key, as eventually we'll try to migrate to go1.17 generics and Delete is one of the most similar methods everywhere.

Reply:
done